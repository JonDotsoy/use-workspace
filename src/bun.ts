import type { Workspace } from "./use-workspace";

export class Bun {
  constructor(readonly workspace: Workspace) {}

  async init() {
    await this.workspace.exec({ cmd: ["bun", "-y", "init", "."] });
  }
}

export const useBun = async (workspace: Workspace) => {
  const bun = new Bun(workspace);

  await bun.init();

  return bun;
};
