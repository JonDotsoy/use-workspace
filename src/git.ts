import type { Workspace } from "./use-workspace.js";
import * as fs from "fs/promises";

type GitConfigOptions = {
  local?: boolean;
};

export class GitConfig {
  constructor(readonly git: Git) {}

  private toFlagLocal(options?: GitConfigOptions) {
    const local = options?.local ?? true;
    if (local) return ["--local"];
    return [];
  }

  async set(
    name: `${string}.${string}`,
    value: string,
    options?: GitConfigOptions,
  ) {
    await this.git.workspace.exec({
      cmd: ["git", "config", ...this.toFlagLocal(), name, value],
    });
  }

  async get(name: `${string}.${string}`, options?: GitConfigOptions) {
    const childProcess = await this.git.workspace.exec({
      cmd: ["git", "config", ...this.toFlagLocal(), "-z", name],
    });
    return await childProcess.toText();
  }
}

export class Git {
  readonly config: GitConfig = new GitConfig(this);

  constructor(readonly workspace: Workspace) {}

  async init() {
    await this.workspace.exec({ cmd: ["git", "init"] });
  }
}

export const useGit = async (workspace: Workspace) => {
  const gitLocation = new URL(".git/", workspace.location);
  const gitExists = await fs.exists(gitLocation);

  const git = new Git(workspace);

  if (!gitExists) {
    await git.init();
  }

  return git;
};
