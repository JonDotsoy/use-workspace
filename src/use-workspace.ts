import { URL } from "url";
import * as fs from "fs/promises";
import { exec } from "./utils/exec.js";

async function* eachNodeModulesFolders() {
  const maxIntents = 200;
  let index = 0;
  let currentAlternative = new URL("./", `file://${__filename}`);
  let isEnd = false;
  while (true) {
    index = index + 1;
    if (index > maxIntents) return;
    const alternativeNodeModule = new URL("node_modules/", currentAlternative);
    const exists = await fs.exists(alternativeNodeModule);
    const stat = exists ? await fs.stat(alternativeNodeModule) : null;
    const isDirectory = stat?.isDirectory() ?? false;
    if (isDirectory) yield alternativeNodeModule;
    currentAlternative = new URL("../", currentAlternative);
    if (isEnd) return;
    if (currentAlternative.pathname === "/") {
      isEnd = true;
    }
  }
}

async function getNodeModulesFolder() {
  for await (const alternative of eachNodeModulesFolders()) {
    return alternative;
  }
  return null;
}

function getWorkspaceLocationByEnv() {
  const WORKSPACE_LOCATION = process.env.WORKSPACE_LOCATION ?? null;
  if (!WORKSPACE_LOCATION) return null;
  const cwd = `file://${process.cwd()}/`;
  if (!URL.canParse(WORKSPACE_LOCATION, cwd)) return null;
  return new URL(WORKSPACE_LOCATION, cwd);
}

async function getBaseWorkspaceLocation() {
  const byEnv = getWorkspaceLocationByEnv();
  if (byEnv) return byEnv;
  const nodeModules = await getNodeModulesFolder();
  if (!nodeModules)
    throw new Error(
      `Missing node_modules folder. Please describe the environment WORKSPACE_LOCATION.`,
    );
  return new URL(".workspaces/", nodeModules);
}

const toWorkspaceLocation = async (relativeName: string | URL) => {
  if (relativeName instanceof URL) return relativeName;
  return new URL(`${relativeName}/`, await getBaseWorkspaceLocation());
};

type WorkspaceOptions = {
  /** Clean all content of workspace if exists */
  cleanBefore?: boolean;
  /** URL format */
  template?: string;
};

export class Workspace {
  constructor(readonly location: URL) {}

  async exec(options: { cmd: string[]; silent?: boolean }) {
    return await exec({
      cmd: options.cmd,
      cwd: this.location,
      silent: options.silent,
    });
  }
}

export const useWorkspace = async (
  relativeName: string | URL = "default",
  options?: WorkspaceOptions,
): Promise<Workspace> => {
  const cleanBefore = options?.cleanBefore ?? false;
  const template = options?.template ? new URL(options?.template) : null;

  const workspaceLocation = await toWorkspaceLocation(relativeName);

  if (cleanBefore) await fs.rm(workspaceLocation, { recursive: true });

  await fs.mkdir(workspaceLocation, { recursive: true });

  // Copy the template on the workspace
  if (template) {
    const templateFiles = await fs.readdir(template, {
      recursive: true,
      withFileTypes: true,
    });
    for (const direct of templateFiles) {
      if (direct.isFile()) {
        const source = new URL(direct.name, template);
        const target = new URL(direct.name, workspaceLocation);
        await fs.mkdir(new URL("./", source), { recursive: true });
        await fs.copyFile(source, target);
      }
    }
  }

  return new Workspace(workspaceLocation);
};
