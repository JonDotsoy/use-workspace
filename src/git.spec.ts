import { test, expect } from "bun:test";
import { useWorkspace } from "./use-workspace.js";
import { useGit } from "./git.js";
import fs from "fs/promises";

test("should call useGit expected the .git folder (workspace foo)", async () => {
  const workspace = await useWorkspace("foo", { cleanBefore: true });
  await useGit(workspace);

  expect(await fs.readdir(workspace.location)).toMatchSnapshot(
    "list workspace",
  );
});

test("should call useGit expected the .git folder (workspace foo2)", async () => {
  const workspace = await useWorkspace("foo2");
  const git = await useGit(workspace);

  await git.config.set("aa.bb", "cc");
  await git.config.set("aa.dd", "ee\n");

  expect(await git.config.get("aa.bb")).toEqual("cc");
  expect(await git.config.get("aa.dd")).toEqual("ee\n");

  expect(
    await fs.readFile(new URL(".git/config", workspace.location), "utf-8"),
  ).toMatchSnapshot("list workspace");
});
