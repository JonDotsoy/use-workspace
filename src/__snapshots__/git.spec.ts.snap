// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should call useGit expected the .git folder (workspace foo): list workspace 1`] = `
[
  ".git",
]
`;

exports[`should call useGit expected the .git folder (workspace foo2): list workspace 1`] = `
"[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
	ignorecase = true
	precomposeunicode = true
[aa]
	bb = cc
	dd = ee\n
"
`;
